### PR TITLE
expression: scalar function update collacte #35149

### LINKED ISSUE #35149: Different  collation cause some execution plans to be unavailable
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
 CREATE TABLE `A` (
  `CST_ID` char(18) NOT NULL DEFAULT ' ',
  `RCRD_EXPY_TMS` datetime(3) NOT NULL ,
  `MULTI_TENANCY_ID` char(5) NOT NULL DEFAULT ' ',
  `IDV_LGL_NM` varchar(180) NOT NULL DEFAULT ' ' ,
  `IDV_NM_CPA_FULLNM` varchar(60) NOT NULL DEFAULT ' ' ,
  `CRDT_TPCD` char(4) NOT NULL DEFAULT ' ' ,
  `CRDT_NO` varchar(120) NOT NULL DEFAULT ' ' ,
  `BRTH_DT` char(8) NOT NULL DEFAULT ' ' ,
  `GND_CD` char(2) NOT NULL DEFAULT ' ' ,
  `NAT_CD` char(3) NOT NULL DEFAULT ' ' ,
  `RSDNC_NAT_CD` char(3) NOT NULL DEFAULT ' ' ,
  `PREF_LNG_CD` char(2) NOT NULL DEFAULT ' ' ,
  `HSHLDRGST_ADIV_CD` char(6) NOT NULL DEFAULT ' ' ,
  `MAR_STTN_CD` char(2) NOT NULL DEFAULT ' ' ,
  `CHL_STTN_CD` char(2) NOT NULL DEFAULT ' ' ,
  `ETHNCT_CD` char(2) NOT NULL DEFAULT ' ' ,
  `RLG_CD` char(4) NOT NULL DEFAULT ' ' ,
  `PLTCLPARTY_CD` char(2) NOT NULL DEFAULT ' ' ,
  `LCS_CD` char(6) NOT NULL DEFAULT ' ' ,
  `BLNG_INSID` char(9) NOT NULL DEFAULT ' ' ,
  `CCB_EMPE_IND` char(1) NOT NULL DEFAULT ' ',
  `PLN_FNC_EFCT_IND` char(1) NOT NULL DEFAULT ' ',
  `IMPT_PSNG_IND` char(1) NOT NULL DEFAULT ' ' ,
  `PTNL_VIP_IND` char(1) NOT NULL DEFAULT ' ' ,
  `SPCLVIP_IND` char(1) NOT NULL DEFAULT ' ' ,
  `STM_EVL_CST_GRD_CD` char(4) NOT NULL DEFAULT ' ',
  `MNUL_EVL_CST_GRD_CD` char(4) NOT NULL DEFAULT ' ' ,
  `PRVT_BNK_CST_GRD_CD` char(4) NOT NULL DEFAULT ' ' ,
  `PRVT_BNK_SIGN_CST_IND` char(1) NOT NULL DEFAULT ' ',
  `MO_INCMAM` decimal(15,2) NOT NULL DEFAULT '0',
  `CSTMGR_ID` varchar(12) NOT NULL DEFAULT ' ',
  `BEST_CTC_TM_CD` char(2) NOT NULL DEFAULT ' ' ,
  `BEST_CTC_MTDCD` char(2) NOT NULL DEFAULT ' ' ,
  `SPNMLT_IND` char(1) NOT NULL DEFAULT ' ' ,
  `RELPARTY_IND` char(1) NOT NULL DEFAULT ' ' ,
  `REG_AFM_RELPARTY_IND` char(1) NOT NULL DEFAULT ' ',
  `TEL_TPCD` char(2) NOT NULL DEFAULT ' ' ,
  `BEST_CTC_TEL` varchar(50) NOT NULL DEFAULT ' ',
  `LV1_BR_NO` char(9) NOT NULL DEFAULT ' ',
  `PREF_MSND_MTDCD` char(2) NOT NULL DEFAULT ' ' ,
  `RCV_MAIL_ADR_TPCD` char(2) NOT NULL DEFAULT ' ' ,
  `ENTP_ADV_MGTPPL_IND` char(1) NOT NULL DEFAULT ' ' ,
  `ENTP_ACT_CTRL_PSN_IND` char(1) NOT NULL DEFAULT ' ' ,
  `ENLGPS_IND` char(1) NOT NULL DEFAULT ' ' ,
  `CST_CHNL_BSOP_IND` char(1) NOT NULL DEFAULT ' ' ,
  `EMPCHNL_BSOP_IND` char(1) NOT NULL DEFAULT ' ' ,
  `IDCST_AUM_BAL` decimal(15,2) NOT NULL DEFAULT '0' ,
  `MT_ENT_IND` char(1) NOT NULL DEFAULT ' ' ,
  `NON_RSDNT_IND` char(1) NOT NULL DEFAULT ' ' ,
  `WRK_UNIT_CHAR_CD` char(4) NOT NULL DEFAULT ' ' ,
  `WRK_UNIT_NM` varchar(600) NOT NULL DEFAULT ' ',
  `RSRV_FLD1_INF` varchar(60) NOT NULL DEFAULT ' ' ,
  `RSRV_FLD2_INF` varchar(60) NOT NULL DEFAULT ' ' ,
  `RSRV_FLD3_INF` varchar(60) NOT NULL DEFAULT ' ',
  `CRT_INSID` char(9) NOT NULL DEFAULT ' ',
  `CRT_EMPID` char(8) NOT NULL DEFAULT ' ' ,
  `LAST_UDT_INSID` char(9) NOT NULL DEFAULT ' ' ,
  `LAST_UDT_EMPID` char(8) NOT NULL DEFAULT ' ' ,
  `CUR_STM_CRT_TMS` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3) ,
  `CUR_STM_UDT_TMS` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3) ,
  `LCL_YRMO_DAY` char(8) NOT NULL DEFAULT ' ' ,
  `LCL_HR_GRD_SCND` char(9) NOT NULL DEFAULT ' ' ,
  `CRT_STM_NO` char(6) NOT NULL DEFAULT ' ' ,
  `UDT_STM_NO` char(6) DEFAULT ' ' ,
  `SRCSYS_CRT_TMS` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3) ,
  `SRCSYS_UDT_TMS` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3) ,
  PRIMARY KEY (`CST_ID`,`RCRD_EXPY_TMS`,`MULTI_TENANCY_ID`) /*T![clustered_index] CLUSTERED */,
  KEY `TXPC10101` (`IDV_LGL_NM`,`BRTH_DT`),
  KEY `TXPC10102` (`IDV_NM_CPA_FULLNM`,`BRTH_DT`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;

CREATE TABLE `B` (
  `MULTI_TENANCY_ID` char(5) NOT NULL DEFAULT ' ',
  `CST_ID` char(18) NOT NULL DEFAULT ' ',
  `STAT_MO` char(6) NOT NULL DEFAULT ' ',
  `IDCST_AUM_BAL` decimal(15,2) NOT NULL DEFAULT '0',
  `IDCST_AUM_JAN_BAL` decimal(15,2) NOT NULL DEFAULT '0',
  `IDCST_AUM_FEB_BAL` decimal(15,2) NOT NULL DEFAULT '0',
  `IDCST_AUM_MAR_BAL` decimal(15,2) NOT NULL DEFAULT '0',
  `IDCST_AUM_APR_BAL` decimal(15,2) NOT NULL DEFAULT '0',
  `IDCST_AUM_MAY_BAL` decimal(15,2) NOT NULL DEFAULT '0',
  `IDCST_AUM_JUN_BAL` decimal(15,2) NOT NULL DEFAULT '0',
  `IDCST_AUM_JUL_BAL` decimal(15,2) NOT NULL DEFAULT '0',
  `IDCST_AUM_AUG_BAL` decimal(15,2) NOT NULL DEFAULT '0',
  `IDCST_AUM_SEPT_BAL` decimal(15,2) NOT NULL DEFAULT '0',
  `IDCST_AUM_OCT_BAL` decimal(15,2) NOT NULL DEFAULT '0',
  `IDCST_AUM_NOV_BAL` decimal(15,2) NOT NULL DEFAULT '0',
  `IDCST_AUM_DEC_BAL` decimal(15,2) NOT NULL DEFAULT '0',
  `IDCSTRCTHRMOCRNACCINT` decimal(15,2) NOT NULL DEFAULT '0',
  `IDCSTRT12MOCRNACC_INT` decimal(15,2) NOT NULL DEFAULT '0',
  PRIMARY KEY (`CST_ID`,`MULTI_TENANCY_ID`) /*T![clustered_index] CLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;

SET NAMES utf8mb4 COLLATE utf8mb4_general_ci;
explain   select  /*+ INL_JOIN(a, b) */     A.*,     CAST(IFNULL(A.Mo_IncmAm,0) as DECIMAL(18,2)) as "Mo_IncmAm",           CAST(IFNULL(B.IdCst_AUM_Bal,0) as DECIMAL(18,2)) as "IdCst_AUM_Bal",         LEFT(DATE_FORMAT(A.SrcSys_Crt_Tms,"%Y%m%d%H%i%S%f"),17) as "SrcSys_Crt_Tms"         from A left join B on A.Cst_ID=B.Cst_ID and A.MULTI_TENANCY_ID=B.MULTI_TENANCY_ID where         A.MULTI_TENANCY_ID="CN000" and         A.Cst_ID="151644656484" and         A.RCRD_EXPY_TMS=STR_TO_DATE('99991231235959999','%Y%m%d%H%i%S%f');
SET NAMES utf8mb4 COLLATE utf8mb4_bin;
explain   select  /*+ INL_JOIN(a, b) */     A.*,     CAST(IFNULL(A.Mo_IncmAm,0) as DECIMAL(18,2)) as "Mo_IncmAm",           CAST(IFNULL(B.IdCst_AUM_Bal,0) as DECIMAL(18,2)) as "IdCst_AUM_Bal",         LEFT(DATE_FORMAT(A.SrcSys_Crt_Tms,"%Y%m%d%H%i%S%f"),17) as "SrcSys_Crt_Tms"         from A left join B on A.Cst_ID=B.Cst_ID and A.MULTI_TENANCY_ID=B.MULTI_TENANCY_ID where         A.MULTI_TENANCY_ID="CN000" and         A.Cst_ID="151644656484" and         A.RCRD_EXPY_TMS=STR_TO_DATE('99991231235959999','%Y%m%d%H%i%S%f');

### 2. What did you expect to see? (Required)
Consistent execution plan when using hint
### 3. What did you see instead (Required)
<img width="1916" alt="image" src="https://user-images.githubusercontent.com/97376271/172080022-6bd14001-d131-4a9e-928f-6c18edce7f9d.png">
<img width="1907" alt="image" src="https://user-images.githubusercontent.com/97376271/172080049-f7b3d259-63dc-45da-879f-c0aa9275ebc7.png">
<img width="1902" alt="image" src="https://user-images.githubusercontent.com/97376271/172080080-8f2a811f-6233-4324-a047-2b519bc0bea7.png">
<img width="1919" alt="image" src="https://user-images.githubusercontent.com/97376271/172080099-be051fb8-fca3-4f8d-8361-f68e5bcc5df1.png">

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
MySQL [test]> SELECT tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                           |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v6.0.0-8-g9258e39-dirty
Edition: Community
Git Commit Hash: 9258e39482be50fce59758ccb9ed0a03c010b225
Git Branch: release-6.0
UTC Build Time: 2022-06-05 08:54:27
GoVersion: go1.18
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by keeplearning20221**
### What problem does this PR solve?

Issue Number: close #35149

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by keeplearning20221** [comment]
./check

**Comment by keeplearning20221** [comment]
./check unit-test

**Comment by keeplearning20221** [comment]
./check

**Comment by keeplearning20221** [comment]
/idc-jenkins-ci-tidb/unit-test

File: expression/constant.go [hunk 1]
```diff
@@ -450,3 +451,62 @@ func (c *Constant) Coercibility() Coercibility {
...
+
+}
+
+// ExpressionUpdateCollate checks whether the expression can be updated to the new collation.
+func expressionUpdateCollate(eq *ScalarFunction, colIsLeft bool) bool {
+	var col *Column
+	var con *Constant
+	colOk := false
+	conOk := false
+	if colIsLeft {
```
Comment (TonsnakeLin) [inline expression/constant.go:474]: I think you can judge it by you self but not the input paramter "colIsLeft "

File: expression/constant.go [hunk 2]
```diff
@@ -450,3 +451,62 @@ func (c *Constant) Coercibility() Coercibility {
...
 	return c.collationInfo.Coercibility()
 }
+
+// TryUpdateCollate Update collate based on priority #35149
+func TryUpdateCollate(eq *ScalarFunction) {
+
+	if eq.FuncName.L != ast.EQ {
+		return
+	}
+	ok := expressionUpdateCollate(eq, true)
```
Comment (TonsnakeLin) [inline expression/constant.go:461]: you can simplify as one line: expressionUpdateCollate(eq, true) and remove the second pram of expressionUpdateCollate

File: expression/constant.go [hunk 3]
```diff
@@ -450,3 +451,62 @@ func (c *Constant) Coercibility() Coercibility {
 	}
 	return c.collationInfo.Coercibility()
 }
+
+// TryUpdateCollate Update collate based on priority #35149
```
Comment (TonsnakeLin) [inline expression/constant.go:455]: This function is at very low level, could you explain it in summary

File: expression/constant.go (hunk 1 above)
Comment (keeplearning20221) [reply expression/constant.go:474]: I implemented this function imitating validEqualCond Helper

File: expression/constant.go [hunk 4]
```diff
@@ -450,3 +451,62 @@ func (c *Constant) Coercibility() Coercibility {
...
+	if !conOk {
+		return false
+	}
+	if !con.HasCoercibility() {
+		return true
+	}
+
+	if col.GetType().GetCharset() != con.GetType().GetCharset() {
+		// utf8 is a subset of utf8mb4
+		if con.GetType().GetCharset() == "utf8" && col.GetType().GetCharset() == "utf8mb4" {
```
[the remark this answers is not in the document]
Comment (keeplearning20221) [reply expression/constant.go:503]: done

File: expression/constant.go [hunk 5]
```diff
@@ -450,3 +451,62 @@ func (c *Constant) Coercibility() Coercibility {
...
+		con, conOk = eq.GetArgs()[0].(*Constant)
+	}
+	if !conOk {
+		return false
+	}
+	if !con.HasCoercibility() {
+		return true
+	}
+
+	if col.GetType().GetCharset() != con.GetType().GetCharset() {
```
[the remark this answers is not in the document]
Comment (keeplearning20221) [reply expression/constant.go:500]: done

File: expression/constant.go [hunk 6]
```diff
@@ -450,3 +451,62 @@ func (c *Constant) Coercibility() Coercibility {
 	}
 	return c.collationInfo.Coercibility()
 }
+
+// TryUpdateCollate Update collate based on priority #35149
+func TryUpdateCollate(eq *ScalarFunction) {
+
+	if eq.FuncName.L != ast.EQ {
```
[the remark this answers is not in the document]
Comment (keeplearning20221) [reply expression/constant.go]: done

File: expression/constant.go [hunk 7]
```diff
@@ -450,3 +451,62 @@ func (c *Constant) Coercibility() Coercibility {
...
+		return
+	}
+	ok := expressionUpdateCollate(eq, true)
+	if !ok {
+		expressionUpdateCollate(eq, false)
+	}
+
+}
+
+// ExpressionUpdateCollate checks whether the expression can be updated to the new collation.
```
[the remark this answers is not in the document]
Comment (keeplearning20221) [reply expression/constant.go:466]: done

File: expression/constant.go (hunk 2 above)
Comment (keeplearning20221) [reply expression/constant.go]: done

File: executor/explain_test.go [hunk 8]
```diff
@@ -424,3 +424,48 @@ func TestFix29401(t *testing.T) {
...
+	tk.MustExec("set names utf8 collate utf8_general_ci;")
+	tk.MustQuery("explain select /*+ INL_JOIN(t9,t10) */ * from t9 join t10 on t9.id = t10.id and  t10.id = 's12' ;").Check(testkit.RowsWithSep("|",
+		`HashJoin_7|100.00|root| CARTESIAN inner join`, `├─Point_Get_11(Build)|1.00|root|table:t10, clustered index:PRIMARY(id) `, `└─TableReader_10(Probe)|10.00|root| data:TableRangeScan_9`,
+		`  └─TableRangeScan_9|10.00|cop[tikv]|table:t9|range:["s12","s12"], keep order:false, stats:pseudo`))
+	tk.MustExec("set names utf8 collate utf8_bin;")
+	tk.MustQuery("explain select /*+ INL_JOIN(t9,t10) */ * from t9 join t10 on t9.id = t10.id and  t10.id = 's12' ;").Check(testkit.RowsWithSep("|",
+		`HashJoin_7|100.00|root| CARTESIAN inner join`, `├─Point_Get_11(Build)|1.00|root|table:t10, clustered index:PRIMARY(id) `, `└─TableReader_10(Probe)|10.00|root| data:TableRangeScan_9`,
+		`  └─TableRangeScan_9|10.00|cop[tikv]|table:t9|range:["s12","s12"], keep order:false, stats:pseudo`))
+	tk.MustExec("set names utf8mb4 collate utf8mb4_general_ci;")
+	tk.MustQuery("explain select /*+ INL_JOIN(t9,t10) */ * from t9 join t11 on t9.id = t11.id and  t11.id = 's12' ;").Check(testkit.RowsWithSep("|",
```
Comment (tiancaiamao) [inline executor/explain_test.go:467]: The hint should be `INL_JOIN(t9, t11)` ?

File: executor/explain_test.go [hunk 9]
```diff
@@ -424,3 +424,48 @@ func TestFix29401(t *testing.T) {
...
+	  ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin`)
+	tk.MustExec(`CREATE TABLE t11 (
+		id varchar(100) NOT NULL ,
+		id2 int(11) DEFAULT NULL,
+		PRIMARY KEY (id) /*T![clustered_index] CLUSTERED */
+	  ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin`)
+	tk.MustExec("set names utf8mb4 collate utf8mb4_general_ci;")
+	tk.MustQuery("explain select /*+ INL_JOIN(t9,t10) */ * from t9 join t10 on t9.id = t10.id and  t10.id = 's12' ;").Check(testkit.RowsWithSep("|",
+		`HashJoin_7|100.00|root| CARTESIAN inner join`, `├─Point_Get_11(Build)|1.00|root|table:t10, clustered index:PRIMARY(id) `, `└─TableReader_10(Probe)|10.00|root| data:TableRangeScan_9`,
+		`  └─TableRangeScan_9|10.00|cop[tikv]|table:t9|range:["s12","s12"], keep order:false, stats:pseudo`))
```
Comment (tiancaiamao) [inline executor/explain_test.go:453]: When create table collation is `_bin`, and the session collate is `_general_ci`, `INL_JOIN` join hint can't work?

File: executor/explain_test.go [hunk 10]
```diff
@@ -424,3 +424,48 @@ func TestFix29401(t *testing.T) {
...
+	tk.MustExec("set names utf8mb4 collate utf8mb4_bin;")
+	tk.MustQuery("explain select /*+ INL_JOIN(t9,t10) */ * from t9 join t10 on t9.id = t10.id and  t10.id = 's12' ;").Check(testkit.RowsWithSep("|",
+		`HashJoin_7|100.00|root| CARTESIAN inner join`, `├─Point_Get_11(Build)|1.00|root|table:t10, clustered index:PRIMARY(id) `, `└─TableReader_10(Probe)|10.00|root| data:TableRangeScan_9`,
```
Comment (tiancaiamao) [inline executor/explain_test.go:456]: Why this is a `hash join` instead of `index join`?

File: expression/constant.go [hunk 11]
```diff
@@ -450,3 +451,66 @@ func (c *Constant) Coercibility() Coercibility {
 	}
 	return c.collationInfo.Coercibility()
 }
+
+// TryUpdateEQFunctionCollate Update collate based on priority #35149
+// Try to unify collate by priority before physical optimization.
+// After unifying collate, the impact of session level on some optimizations can be shielded
+func TryUpdateEQFunctionCollate(eq *ScalarFunction) {
```
Comment (tiancaiamao) [inline expression/constant.go:458]: If this function is not used outside the `expression` package, please use lowercase for the function name.

File: expression/constant.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply expression/constant.go:466]: s/ExpressionUpdateCollate/expressionUpdateCollate

File: expression/constant.go [hunk 12]
```diff
@@ -450,3 +451,66 @@ func (c *Constant) Coercibility() Coercibility {
...
+
+	// In addition to utf8 and utf8mb4 encoding, other encodings are handled by the cast function
+	if col.GetType().GetCharset() != con.GetType().GetCharset() {
+		// utf8 is a subset of utf8mb4
+		// If the constant is utf8 and the column is utf8mb4, the constant can be directly converted to utf8mb4
+		if con.GetType().GetCharset() == "utf8" && col.GetType().GetCharset() == "utf8mb4" {
+			// The lower the coer value, the higher the priority
+			if col.Coercibility() < con.Coercibility() {
+				con.GetType().SetCharset(col.GetType().GetCharset())
+				con.GetType().SetCollate(col.GetType().GetCollate())
```
Comment (tiancaiamao) [inline expression/constant.go:507]: Modify the expression may cause two problems:
1. data race
2. some place may referencing the expression and find  it's unexpectedly modified ... this kind is difficult to debug

"X eq Y" and after the expression execution, they are modified ... this is a footgun, you're playing with fire.

**Comment by tiancaiamao** [comment]
PTAL @wjhuang2016

File: executor/explain_test.go (hunk 8 above)
Comment (tiancaiamao) [reply executor/explain_test.go:467]: Because you're joining t9 and t11

File: executor/explain_test.go (hunk 9 above)
Comment (keeplearning20221) [reply executor/explain_test.go:453]: In scenarios where pushdown is possible, hash join will be used

File: expression/constant.go (hunk 11 above)
Comment (keeplearning20221) [reply expression/constant.go:458]: OK

File: executor/explain_test.go (hunk 10 above)
Comment (keeplearning20221) [reply executor/explain_test.go:456]: The optimizer cannot build an index join

**Pull request closed without merging**